CRM/Utils/Type.php

```diff
@@ -77,70 +77,13 @@ class CRM_Utils_Type {
    *   String identifying the data type, e.g. 'Int' or 'String'.
    */
   public static function typeToString($type) {
-    // @todo Use constants in the case statements, e.g. "case T_INT:".
-    // @todo return directly, instead of assigning a value.
-    // @todo Use a lookup array, as a property or as a local variable.
-    switch ($type) {
-      case 1:
-        $string = 'Int';
-        break;
-
-      case 2:
-        $string = 'String';
-        break;
-
-      case 3:
-        $string = 'Enum';
-        break;
-
-      case 4:
-        $string = 'Date';
-        break;
-
-      case 8:
-        $string = 'Time';
-        break;
-
-      case 16:
-        $string = 'Boolean';
-        break;
-
-      case 32:
-        $string = 'Text';
-        break;
-
-      case 64:
-        $string = 'Blob';
-        break;
-
-      // CRM-10404
-      case 12:
-      case 256:
-        $string = 'Timestamp';
-        break;
-
-      case 512:
-        $string = 'Float';
-        break;
-
-      case 1024:
-        $string = 'Money';
-        break;
-
-      case 2048:
-        $string = 'Date';
-        break;
-
-      case 4096:
-        $string = 'Email';
-        break;
-
-      case 16384:
-        $string = 'Mediumblob';
-        break;
+    // CRM-10404
+    if ($type == 12) {
+      return 'Timestamp';
     }
 
-    return (isset($string)) ? $string : "";
+    $typeStrings = self::getValidTypes();
+    return (array_search($type, $typeStrings)) ?: '';
```

Comment:
Array key lookups would be faster than `array_search()`. And the function call to ::getValidTypes() could  be avoided if there was a const `CRM_Utils_Type::TYPE_STRINGS = array(self::T_INT => 'Int', ...)`.

You could introduce `CRM_Utils_Type::TYPE_STRINGS` and `CRM_Utils_Type::TYPES_BY_STRING`, with `CRM_Utils_Type::TYPE_STRINGS === array_flip(CRM_Utils_Type::TYPES_BY_STRING)`.

Not sure if "string" is really a good way to say it, maybe "label" or "machine name"? Depends how this is used. So then it would be TYPES_BY_LABEL or TYPES_BY_NAME, or whatever.

I think in some earlier PHP versions, array-valued const were not possible, but now they are. Check 3v4l.org.

Reply:
The main goal here is to reduce the number of times all the types are listed out because that will reduce the opportunity for error (like the whole problem of `2048` being `Email` or `Date`).
I could do an `array_flip()` and then look by key, but doing a quick experiment of that at 3v4l.org (which is awesome BTW), that approach appears to take the same time and memory on PHP 7.0 and slightly higher on 5.6, 7.1, and 7.2.

Reply:
If you repeat the array_flip() on each call then yes it won't make anything faster.
Some ideas about this:
- Hardcode both lists, and add a unit test which verifies the array_flip() equality.
- Build both direction of the list with define(), outside the class, where array_flip() is possible. Then assign the defined constants to class constants. This only works in PHP 7 apparently, https://3v4l.org/er4na
- Lazy-initialize a local static variable in CRM_Utils_Type::getValidTypes() using array_flip().

I am not saying that either of those is perfectly desirable, just saying the options that I can think of.

Reply:
I did a quick grep & didn't find any places this function is called where I would expect it to be hammered so I wouldn't put too much effort into performance at this level